fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java

```diff
@@ -192,44 +194,28 @@ public static FanType valueOf(RecordMetaDataProto.Field.FanType fanType) throws
         return Collections.singletonList(this);
     }
 
-    /**
-     * Flatten this key expression into a list of {@link Element}s, much like {@link #normalizeKeyForPositions()}.
-     * By default, this method throws an exception because most key expressions cannot be flattened to a list of
-     * elements without prior adjustment. This method is only overriden by key expressions that can be flattened.
-     * @return a list of elements representing this key expression in unnested form
-     * @see ElementKeyExpression#flattenForPlanner()
-     * @see ThenKeyExpression#flattenForPlanner()
-     */
     @API(API.Status.EXPERIMENTAL)
     @Nonnull
-    default List<Element> flattenForPlanner() {
-        throw new RecordCoreException("illegal non-element expression");
+    default ExpandedPredicates normalizeForPlanner(@Nonnull CorrelationIdentifier baseAlias, @Nonnull final Supplier<CorrelationIdentifier> parameterAliasSupplier) {
+        return normalizeForPlanner(baseAlias, parameterAliasSupplier, Collections.emptyList());
     }
 
+    @API(API.Status.EXPERIMENTAL)
+    @Nonnull
+    ExpandedPredicates normalizeForPlanner(@Nonnull CorrelationIdentifier baseAlias, @Nonnull final Supplier<CorrelationIdentifier> parameterAliasSupplier, @Nonnull List<String> fieldNamePrefix);
```

Comment:
I have some conceptual difficulty with this. I think of the normalized form of a `KeyExpression` mostly as one that unnests while adding correlation names, so that an ordered sequence of values can be matched with the proper unification semantics. So it is surprising that the normal forms of keys and queries are of the same stuff.

Perhaps what is strange is that a `QueryPredicate` can have a placeholder. A `ValueComparisonRangePredicate` is not so much a predicate as a thing to which predicates can be added as part of that matching.

It may be that the problem stems primarily from naming.

Reply:
Your observations are correct, a placeholder is a predicate in a sense that it only is meaningful in conjunction with an actual match. Maybe another way of thinking about it is to say a place holder is `x in [-inf, inf]` where the matching yields a compensation that is the actual range. It is possible to do it like so but I feel like it may be a bit to convoluted as such a predicate needs to be understood as something the materialized view can handle, it needs special logic to map the range back to an ordinal position in the index, etc. So rather than using `Compensation`s to capture this mapping I opted to bypass and to just have a mapping from query to placeholder which directly gives me the parameter alias, etc. Hope this makes sense.